src/parser/mod.rs

```diff
@@ -9528,61 +9541,81 @@ impl<'a> Parser<'a> {
-                        let values = self.parse_values(is_mysql)?;
-                        MergeClause::NotMatched {
-                            predicate,
-                            columns,
-                            values,
-                        }
-                    }
-                    Some(_) => {
-                        return Err(ParserError::ParserError(
-                            "expected UPDATE, DELETE or INSERT in merge clause".to_string(),
-                        ));
+                }
+                Some(Keyword::DELETE) => {
+                    if matches!(
+                        clause_type,
+                        MergeClauseType::NotMatched | MergeClauseType::NotMatchedByTarget
+                    ) {
+                        return Err(ParserError::ParserError(format!(
+                            "DELETE is not allowed in a {clause_type} merge clause"
+                        )));
                     }
-                    None => {
-                        return Err(ParserError::ParserError(
-                            "expected UPDATE, DELETE or INSERT in merge clause".to_string(),
-                        ));
+                    MergeAction::Delete
+                }
+                Some(Keyword::INSERT) => {
+                    if !matches!(
+                        clause_type,
+                        MergeClauseType::NotMatched | MergeClauseType::NotMatchedByTarget
+                    ) {
+                        return Err(ParserError::ParserError(format!(
+                            "INSERT is not allowed in a {clause_type} merge clause"
+                        )));
                     }
-                },
-            );
+                    let is_mysql = dialect_of!(self is MySqlDialect);
+
+                    let insert_expr = if dialect_of!(self is BigQueryDialect | GenericDialect)
+                        && self.parse_keyword(Keyword::ROW)
+                    {
+                        MergeInsertExpr::Row
+                    } else {
+                        let columns = self.parse_parenthesized_column_list(Optional, is_mysql)?;
+                        if dialect_of!(self is BigQueryDialect | GenericDialect)
+                            && self.parse_keyword(Keyword::ROW)
+                        {
+                            MergeInsertExpr::RowWithColumns { columns }
+                        } else {
+                            self.expect_keyword(Keyword::VALUES)?;
+                            let values = self.parse_values(is_mysql)?;
+                            MergeInsertExpr::Values { columns, values }
+                        }
+                    };
+                    MergeAction::Insert(insert_expr)
+                }
+                Some(_) => {
+                    return Err(ParserError::ParserError(
+                        "expected UPDATE, DELETE or INSERT in merge clause".to_string(),
+                    ));
+                }
```

Comment:
```suggestion
                Some(_) => unreachable!(),
```

Reply:
Updated, I've merged the `Some(_) | None` clause, it still returns an error, I figure that's preferred here vs panicking in case a mistake slips in